### PR TITLE
41 type information

### LINKED ISSUE #41: Provide valuable type information
Type definitions serve multiple purposes. One one side, they help the compiler to determine choices depending of context and enrich the development workflow.   
On the other side, do they provide documentation for possible consumers and readers.

In terms of  #15 and #30, I would like to suggest to enrich the project with valuable type information to promote the usage of this library.

### PR DESCRIPTION
**Pull request by Retro64**
close #41

### CODE REVIEW

**Comment by Retro64** [comment]
:+1:

File: README.md [hunk 1]
```diff
@@ -30,7 +30,7 @@ import {
 } from 'lib-oauth-tooling';
 ```
 
-#### TokenCache(tokenConfig: any, oauthConfig: any)
+#### TokenCache(tokenConfig: { [key : string]: string[] }, oauthConfig: OAuthConfig)
```
Comment (ISO50) [inline README.md]: Question: should we move the signature out of the headline?

File: integration-test/mock-tooling/index.ts [hunk 2]
```diff
@@ -83,17 +83,17 @@ describe('mock tooling', () => {
 
       // when
       let promise = getTokenInfo(tokeninfoEndpoint, 'foo')
-        .then((token: any) => {
+        .then((token: TokenInfo) => {
```
Comment (ISO50) [inline integration-test/mock-tooling/index.ts]: const + indent

File: integration-test/oauth-tooling/getAccessToken.spec.ts [hunk 3]
```diff
@@ -199,7 +199,7 @@ describe('getAccessToken', () => {
       const responseObject = { 'access_token': '4b70510f-be1d-4f0f-b4cb-edbca2c79d41' };
 
       nock(host)
-        .post('/access_token?realm=/services', (body) => {
+        .post('/access_token?realm=/services', (body: any) => {
```
Comment (ISO50) [inline integration-test/oauth-tooling/getAccessToken.spec.ts]: indent

File: package.json [hunk 4]
```diff
@@ -19,28 +19,36 @@
...
     "package.json"
   ],
-  "engines" : {
-    "node" : ">=6.0"
+  "engines": {
+    "node": ">=6.0"
   },
   "dependencies": {
-    "body-parser": "^1.16.0",
+    "body-parser": "^1.17.1",
```
Comment (ISO50) [inline package.json]: Should we switch to non ^ ~ versions for the library?

File: src/TokenCache.ts [hunk 5]
```diff
@@ -57,8 +52,8 @@ class TokenCache {
    * @param {string} The key configured on the tokenCache instance
    * @return {Promise<string>} the resolved access_token
    */
-  public resolveAccessTokenFactory(key: string): () => Promise<string> {
-    return () => this
+  public resolveAccessTokenFactory(key : string) : () => Promise < string > {
```
Comment (ISO50) [inline src/TokenCache.ts]: `<string>`

File: src/TokenCache.ts [hunk 6]
```diff
@@ -57,8 +52,8 @@ class TokenCache {
...
+    return() => this
```
Comment (ISO50) [inline src/TokenCache.ts]: return ()

File: src/TokenCache.ts [hunk 7]
```diff
@@ -70,36 +65,32 @@ class TokenCache {
    * @param tokenName
    * @returns {Promise<T>}
    */
-  get(tokenName: string): Promise<any> {
+  get(tokenName : string) : Promise <TokenInfo> {
```
Comment (ISO50) [inline src/TokenCache.ts]: `get(tokenName: string): Promise<TokenInfo>`

File: src/TokenCache.ts [hunk 8]
```diff
@@ -70,36 +65,32 @@ class TokenCache {
...
 
-          return getAccessToken(config)
-            .then((token: any) => {
+          return getAccessToken(config).then((token : Token) => {
 
-              return getTokenInfo(this.oauthConfig.tokenInfoEndpoint, token.access_token);
-            })
-            .then((tokenInfo: any) => {
+            return getTokenInfo(this.oauthConfig.tokenInfoEndpoint, token.access_token);
+          }).then((tokenInfo : TokenInfo) => {
```
Comment (ISO50) [inline src/TokenCache.ts]:
```
})
.then

```

File: src/TokenCache.ts [hunk 9]
```diff
@@ -112,9 +103,9 @@ class TokenCache {
    * Otherwise, rejects.
    *
    * @param tokenName
-   * @returns {Promise<any>}
+   * @returns {Promise<TokenInfo>}
    */
-  refreshToken(tokenName: string): Promise<any> {
+  refreshToken(tokenName : string) : Promise <TokenInfo> {
```
Comment (ISO50) [inline src/TokenCache.ts]: spaces

File: src/TokenCache.ts [hunk 10]
```diff
@@ -127,13 +118,17 @@ class TokenCache {
    *
    * @returns {Promise<T>}
    */
-  refreshAllTokens(): Promise<any> {
+refreshAllTokens() : Promise <{ [key : string]: TokenInfo }> {
```
Comment (ISO50) [inline src/TokenCache.ts]: spaces

File: src/TokenCache.ts [hunk 11]
```diff
@@ -144,7 +139,7 @@ class TokenCache {
    * @param tokenName
    * @returns {Promise<T>}
    */
-  private validateToken(tokenName: string): Promise<any> {
+  private validateToken(tokenName : string) : Promise <TokenInfo> {
```
Comment (ISO50) [inline src/TokenCache.ts]: spaces

File: src/TokenCache.ts [hunk 12]
```diff
@@ -159,15 +154,12 @@ class TokenCache {
...
+    }).catch(() => {
+      return Promise.reject(`Token ${tokenName} is invalid.`);
+    });
   }
 }
 
-export {
-  TokenCache
-};
+export {TokenCache};
```
Comment (ISO50) [inline src/TokenCache.ts]: Like it was or

```
export { TokenCache };
```

File: src/express-tooling.ts [hunk 13]
```diff
@@ -77,7 +77,7 @@ function handleOAuthRequestMiddleware(options: any) {
   return function(req: any, res: any, next: Function) {
 
     // Skip OAuth validation for paths marked as public
-    if (options.publicEndpoints && match(req.originalUrl, options.publicEndpoints)) {
+if (options.publicEndpoints && match(req.originalUrl, new Set(options.publicEndpoints))) {
```
Comment (ISO50) [inline src/express-tooling.ts]: indent

File: integration-test/mock-tooling/index.ts [hunk 14]
```diff
@@ -82,22 +82,22 @@ describe('mock tooling', () => {
...
+      .then((token: TokenInfo) => {
 
-          expect(token).to.equal(validAuthToken);
-          return getTokenInfo(tokeninfoEndpoint, 'foo');
-        })
-        .then((token : TokenInfo) => {
+        expect(token).to.equal(validAuthToken);
+        return getTokenInfo(tokeninfoEndpoint, 'foo');
+      })
+      .then((token : TokenInfo) => {
```
Comment (ISO50) [inline integration-test/mock-tooling/index.ts]: space

File: src/oauth-tooling.ts [hunk 15]
```diff
@@ -160,45 +160,45 @@ function getAccessToken(options: OAuthConfig): Promise<Token> {
...
-        Object.assign(bodyParameters, {
-          scope: options.scopes.join(' ')
-        });
-      }
+  .then((credentials) => {
+
+    const userData = JSON.parse(credentials[0]);
+    const clientData = JSON.parse(credentials[1]);
+
+    let bodyParameters: any;
```
Comment (ISO50) [inline src/oauth-tooling.ts]: maybe create new issue for new type

**Comment by ISO50** [comment]
👍

**Comment by Retro64** [comment]
:+1:

**Pull request merged**